File: bluetooth_scanner.py


### 1. `bluetooth_scanner.py` (Python)

```python
# bluetooth_scanner.py — Python версия

import asyncio
import json
import csv
import sys
import time
from datetime import datetime
from bleak import BleakScanner
from colorama import init, Fore, Style

init(autoreset=True)

class BluetoothScanner:
    def __init__(self, timeout=5, continuous=False):
        self.timeout = timeout
        self.continuous = continuous
        self.devices = []

    def classify_rssi(self, rssi):
        """Классифицирует качество сигнала по RSSI."""
        if rssi is None:
            return "N/A", Fore.YELLOW
        if rssi >= -50:
            return "Отлично", Fore.GREEN
        elif rssi >= -70:
            return "Средне", Fore.YELLOW
        else:
            return "Плохо", Fore.RED

    async def scan(self):
        """Выполняет сканирование Bluetooth-устройств."""
        print(f"{Fore.CYAN}📶 Bluetooth Scanner (RSSI) (Python)")
        print(f"🔍 Сканирование устройств... (таймаут: {self.timeout} сек)")

        if self.continuous:
            print("🔄 Непрерывный режим (нажмите Ctrl+C для остановки)")
            while True:
                self.devices = []
                devices = await BleakScanner.discover(timeout=self.timeout)
                self.devices = devices
                self.print_table()
                await asyncio.sleep(self.timeout)
        else:
            devices = await BleakScanner.discover(timeout=self.timeout)
            self.devices = devices
            self.print_table()
            self.save_json()
            self.save_csv()

    def print_table(self):
        """Выводит таблицу с устройствами."""
        if not self.devices:
            print(Fore.YELLOW + "❌ Устройства не найдены.")
            return

        # Сортировка по RSSI (от сильного к слабому)
        sorted_devices = sorted(self.devices, key=lambda d: d.rssi if d.rssi else -100, reverse=True)

        print(f"\n{Fore.CYAN}📊 Найденные устройства:{Style.RESET_ALL}")
        print("┌" + "─" * 75 + "┐")
        print(f"│ {'Устройство':<25} {'MAC-адрес':<20} {'RSSI':<10} {'Статус':<15} │")
        print("├" + "─" * 75 + "┤")

        for device in sorted_devices:
            name = device.name or "Неизвестно"
            if len(name) > 25:
                name = name[:22] + "..."
            rssi = device.rssi if device.rssi else 0
            status, color = self.classify_rssi(rssi)
            rssi_str = f"{rssi} dBm" if rssi else "N/A"
            print(f"│ {name:<25} {device.address:<20} {color}{rssi_str:<10}{Style.RESET_ALL} {color}{status:<15}{Style.RESET_ALL} │")

        print("└" + "─" * 75 + "┘")

    def save_json(self, filename="devices.json"):
        """Сохраняет результаты в JSON."""
        data = {
            "timestamp": datetime.now().isoformat(),
            "devices": [
                {
                    "name": d.name or "Неизвестно",
                    "address": d.address,
                    "rssi": d.rssi,
                    "metadata": d.metadata if hasattr(d, 'metadata') else {}
                }
                for d in self.devices
            ]
        }
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"{Fore.GREEN}💾 Сохранено JSON: {filename}")

    def save_csv(self, filename="devices.csv"):
        """Сохраняет результаты в CSV."""
        if not self.devices:
            return
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["Name", "Address", "RSSI"])
            for d in self.devices:
                writer.writerow([d.name or "Неизвестно", d.address, d.rssi])
        print(f"{Fore.GREEN}💾 Сохранено CSV: {filename}")

async def main():
    timeout = 5
    continuous = False

    if len(sys.argv) > 1:
        if sys.argv[1] == "--continuous" or sys.argv[1] == "-c":
            continuous = True
        if len(sys.argv) > 2:
            try:
                timeout = int(sys.argv[2])
            except:
                pass

    scanner = BluetoothScanner(timeout, continuous)
    try:
        await scanner.scan()
    except KeyboardInterrupt:
        print(f"\n{Fore.YELLOW}⏹️ Сканирование остановлено.")
    except Exception as e:
        print(f"{Fore.RED}❌ Ошибка: {e}")

if __name__ == "__main__":
    asyncio.run(main())

File: bluetooth_scanner.js
// bluetooth_scanner.js — JavaScript версия

const noble = require('noble');
const fs = require('fs');
const readline = require('readline');

class BluetoothScanner {
    constructor() {
        this.devices = new Map();
        this.startTime = Date.now();
    }

    classifyRSSI(rssi) {
        if (rssi >= -50) return { status: 'Отлично', color: '\x1b[32m' };
        if (rssi >= -70) return { status: 'Средне', color: '\x1b[33m' };
        return { status: 'Плохо', color: '\x1b[31m' };
    }

    printTable() {
        if (this.devices.size === 0) {
            console.log('\x1b[33m❌ Устройства не найдены.\x1b[0m');
            return;
        }

        // Сортировка по RSSI
        const sorted = Array.from(this.devices.values()).sort((a, b) => b.rssi - a.rssi);

        console.log('\x1b[36m📊 Найденные устройства:\x1b[0m');
        console.log('┌' + '─'.repeat(75) + '┐');
        console.log(`│ ${'Устройство'.padEnd(25)} ${'MAC-адрес'.padEnd(20)} ${'RSSI'.padEnd(10)} ${'Статус'.padEnd(15)} │`);
        console.log('├' + '─'.repeat(75) + '┤');

        for (const d of sorted) {
            const name = d.name || 'Неизвестно';
            const shortName = name.length > 25 ? name.slice(0, 22) + '...' : name;
            const { status, color } = this.classifyRSSI(d.rssi);
            const rssiStr = `${d.rssi} dBm`;
            console.log(`│ ${shortName.padEnd(25)} ${d.address.padEnd(20)} ${color}${rssiStr.padEnd(10)}\x1b[0m ${color}${status.padEnd(15)}\x1b[0m │`);
        }

        console.log('└' + '─'.repeat(75) + '┘');
    }

    saveResults() {
        const data = {
            timestamp: new Date().toISOString(),
            devices: Array.from(this.devices.values()).map(d => ({
                name: d.name || 'Неизвестно',
                address: d.address,
                rssi: d.rssi
            }))
        };

        fs.writeFileSync('devices.json', JSON.stringify(data, null, 2));
        console.log('\x1b[32m💾 Сохранено JSON: devices.json\x1b[0m');

        let csv = 'Name,Address,RSSI\n';
        for (const d of data.devices) {
            csv += `${d.name},${d.address},${d.rssi}\n`;
        }
        fs.writeFileSync('devices.csv', csv);
        console.log('\x1b[32m💾 Сохранено CSV: devices.csv\x1b[0m');
    }

    start() {
        console.log('\x1b[36m📶 Bluetooth Scanner (RSSI) (JavaScript)\x1b[0m');
        console.log('🔍 Сканирование устройств... (нажмите Ctrl+C для остановки)');

        noble.on('stateChange', (state) => {
            if (state === 'poweredOn') {
                noble.startScanning([], true);
            } else {
                console.log(`⚠️ Bluetooth состояние: ${state}`);
            }
        });

        noble.on('discover', (peripheral) => {
            const address = peripheral.address;
            const name = peripheral.advertisement.localName || 'Неизвестно';
            const rssi = peripheral.rssi;

            // Обновляем информацию об устройстве
            this.devices.set(address, { name, address, rssi });

            // Очищаем экран и выводим таблицу
            console.clear();
            this.printTable();
        });

        // Обработка выхода
        process.on('SIGINT', () => {
            noble.stopScanning();
            console.log('\n⏹️ Сканирование остановлено.');
            this.saveResults();
            process.exit(0);
        });
    }
}

// Проверка зависимостей
try {
    require.resolve('noble');
} catch (e) {
    console.error('\x1b[31m❌ Установите noble: npm install noble\x1b[0m');
    process.exit(1);
}

const scanner = new BluetoothScanner();
scanner.start();
